src/Mapping/Driver/AbstractAnnotationDriver.php

```diff
@@ -37,6 +38,10 @@ abstract class AbstractAnnotationDriver implements AnnotationDriverInterface
      */
     public function setAnnotationReader($reader)
     {
+        if (!($reader instanceof Reader)) {
+            trigger_deprecation('gedmo/doctrine-extensions', '3.3', 'Providing an annotation reader which does not implement %s is deprecated, as of 4.0 the reader must implement the interface.', Reader::class);
```

Comment:
I don't know what the opinion of those maintaining or using this library is regarding runtime deprecations, but IMO this is a case where they're needed.

Right now, static analysis isn't great with this library and a combination of static analysis and integration tests wouldn't actually catch the deprecated API usage without fully removing the `object` typehints in the doc blocks and just typehinting the `Doctrine\Common\Annotations\Reader` interface.  So, the added runtime deprecations can help to flag the behavior this PR proposes to deprecate while also giving a meaningful explanation on how to mitigate it.

I went for Symfony's [deprecation contracts](https://github.com/symfony/deprecation-contracts) for the runtime usage as it's what I know personally.  There's also Doctrine's [deprecations package](https://github.com/doctrine/deprecations) which they've been using.

Reply:
I like this 👍🏼